src/services/ui/src/api/submissionService.ts

```diff
@@ -0,0 +1,147 @@
+import { API } from "aws-amplify";
+import {OnemacAttachmentSchema, Authority, ReactQueryApiError} from "shared-types";
+import {SubmissionServiceEndpoint} from "@/lib";
+import {OneMacUser} from "@/api/useGetUser";
+import {useMutation, UseMutationOptions} from "@tanstack/react-query";
+
+export enum SubmissionVariant {
+  INITIAL = "initial",
+  ACTION = "action"
+}
+type SubmissionServiceParameters<T> = {
+  data: T,
+  endpoint: SubmissionServiceEndpoint
+  variant: SubmissionVariant,
+  user: OneMacUser | undefined,
+  authority?: Authority
+}
+type SubmissionServiceResponse = {
+  body: {
+    message: string;
+  };
+};
+type PreSignedURL = {
+  url: string;
+  key: string;
+  bucket: string;
+};
+type UploadRecipe = PreSignedURL & {
+  data: File[],
+  title: string
+}
+type AttachmentsSchema = Record<string, File[] | undefined>
+
+/** Grabs one pre-signed URL for file upload */
+const getPreSignedUrl = async (): Promise<PreSignedURL> => {
+  // TODO: Can this be a GET instead of POST w/ empty body
+  return await API.post("os", "/getUploadUrl", {
+    body: {},
+  });
+};
+
+/** Given an {@link AttachmentSchema} and array of pre signed URLs, this
+ * will return an array of objects that bundle a URL, s3 key, and s3 bucket
+ * with the file(s) to upload */
+const buildUploadRecipes = (
+  attachments: AttachmentsSchema,
+  validAttachmentKeys: string[],
+  preSignedUrls: PreSignedURL[]
+): UploadRecipe[] => {
+  return preSignedUrls.map((obj, idx) => ({
+    ...obj,
+    // Using validAttachmentKeys ensures we don't have any undefined value
+    data: attachments[validAttachmentKeys[idx]] as File[],
+    title: validAttachmentKeys[idx]
+  }));
+};
+
+/** Uploads one file if given a file and its meta from {@link getPreSignedUrl} */
+const uploadAttachment = async (file: File, preSignedURL: PreSignedURL) => {
+  return await fetch(preSignedURL.url, {
+    body: file,
+    method: "PUT",
+  });
+};
+
+/** Pass in an array of UploadRecipes and get a back-end compatible object
+ * to store attachment data */
+const buildAttachmentObject = (
+  recipes: UploadRecipe[]
+): OnemacAttachmentSchema[] => {
+  return recipes.map((r) => r.data.map(f => ({
+    key: r.key,
+    filename: f.name,
+    title: r.title,
+    bucket: r.bucket,
+    uploadDate: Date.now()
+  }) satisfies OnemacAttachmentSchema)).flat();
+};
+
+const buildSubmissionPayload = <T extends Record<string, unknown>>(
+  data: T,
+  variant: SubmissionVariant,
+  user: OneMacUser | undefined,
+  authority?: string,
+  attachments?: UploadRecipe[],
+) => {
+  switch (variant) {
+    case SubmissionVariant.ACTION:
```

Comment:
Could be a good idea to remove the need for a "variant" and just use the given endpoint url to derive payload shape. Couples the idea of the endpoint with a payload shape. That logic can continue when we have variations of initial submissions and new actions. Just one less enum to update!